File: internal/dbipc/host_version.go
package dbipc

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
)

const MinimumHostVersion = "0.10.0"

var semverPattern = regexp.MustCompile(`^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$`)

type initParams struct {
	HostVersion string `json:"host_version"`
}

func ValidateHostVersion(raw json.RawMessage) error {
	var params initParams
	if len(raw) == 0 || string(raw) == "null" {
		return fmt.Errorf("this driver requires Navop >= %s; host_version is missing", MinimumHostVersion)
	}
	if err := json.Unmarshal(raw, &params); err != nil {
		return fmt.Errorf("invalid init params: %w", err)
	}
	major, minor, _, prerelease, ok := parseSemver(params.HostVersion)
	if !ok {
		return fmt.Errorf("this driver requires Navop >= %s; invalid host_version %q", MinimumHostVersion, params.HostVersion)
	}
	if prerelease != "" || (major == 0 && minor < 10) {
		return fmt.Errorf("this driver requires Navop >= %s; current host version is %s. Please upgrade Navop", MinimumHostVersion, params.HostVersion)
	}
	return nil
}

func parseSemver(value string) (major, minor, patch int, prerelease string, ok bool) {
	match := semverPattern.FindStringSubmatch(value)
	if match == nil {
		return 0, 0, 0, "", false
	}
	major64, err1 := strconv.ParseInt(match[1], 10, 32)
	minor64, err2 := strconv.ParseInt(match[2], 10, 32)
	patch64, err3 := strconv.ParseInt(match[3], 10, 32)
	if err1 != nil || err2 != nil || err3 != nil {
		return 0, 0, 0, "", false
	}
	return int(major64), int(minor64), int(patch64), match[4], true
}

File: java/gbase8s-ipc-driver/src/main/java/com/navop/gbase8s/server/HostVersion.java
package com.navop.gbase8s.server;

import java.math.BigInteger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class HostVersion {
    static final String MINIMUM = "0.10.0";
    private static final Pattern SEMVER = Pattern.compile(
            "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(?:-([0-9A-Za-z-]+(?:\\.[0-9A-Za-z-]+)*))?(?:\\+[0-9A-Za-z-]+(?:\\.[0-9A-Za-z-]+)*)?$");

    private HostVersion() {
    }

    static String incompatibility(String version) {
        Matcher matcher = SEMVER.matcher(version == null ? "" : version);
        if (!matcher.matches()) {
            return "this driver requires Navop >= " + MINIMUM + "; invalid or missing host_version `" + version + "`";
        }
        BigInteger major = new BigInteger(matcher.group(1));
        BigInteger minor = new BigInteger(matcher.group(2));
        boolean tooOld = matcher.group(4) != null
                || (major.signum() == 0 && minor.compareTo(BigInteger.TEN) < 0);
        if (tooOld) {
            return "this driver requires Navop >= " + MINIMUM + "; current host version is "
                    + version + ". Please upgrade Navop";
        }
        return null;
    }
}
